src/Features/LanguageServer/ProtocolUnitTests/Miscellaneous/LspMiscellaneousFilesWorkspaceTests.cs

```diff
@@ -0,0 +1,191 @@
+
+        // Open a loose file and verify it gets added to the misc workspace.
+        var looseFileUri = new Uri(@"C:\SomeFile.cs");
+        await testLspServer.OpenDocumentAsync(looseFileUri, source).ConfigureAwait(false);
+        Assert.Equal(looseFileUri.AbsolutePath, GetMiscellaneousDocument(testLspServer)!.FilePath);
+
+        // Verify the loose file is removed from the misc workspace on close.
+        await testLspServer.CloseDocumentAsync(looseFileUri).ConfigureAwait(false);
+        Assert.Null(GetMiscellaneousDocument(testLspServer));
+    }
+
+    [Fact]
+    public async Task TestFileInRegisteredWorkspace_Opened()
+    {
+        var source =
+@"class A
+{
+    void M()
+    {
+    }
+}";
+
+        // Create the server and verify no misc files present.
+        using var testLspServer = CreateTestLspServer(source, out var _);
+        Assert.Null(GetMiscellaneousDocument(testLspServer));
+
+        // Open a file that is part of a registered workspace and verify it is not present in the misc workspace.
+        var fileInWorkspaceUri = new Uri(testLspServer.GetCurrentSolution().Projects.Single().Documents.Single().FilePath);
+        await testLspServer.OpenDocumentAsync(fileInWorkspaceUri).ConfigureAwait(false);
+        Assert.Null(GetMiscellaneousDocument(testLspServer));
+    }
+
+    [Fact]
+    public async Task TestLooseFile_MovedToRegisteredWorkspace()
+    {
+        var source =
+@"class A
+{
+    void M()
+    {
+    }
+}";
+
+        // Create the server and verify no misc files present.
+        using var testLspServer = CreateTestLspServer(string.Empty, out var _);
+        Assert.Null(GetMiscellaneousDocument(testLspServer));
+
+        // Open a loose file and verify it gets added to the misc workspace.
+        var looseFileUri = new Uri(@"C:\SomeFile.cs");
+        await testLspServer.OpenDocumentAsync(looseFileUri, source).ConfigureAwait(false);
+        Assert.Equal(looseFileUri.AbsolutePath, GetMiscellaneousDocument(testLspServer)!.FilePath);
+
+        // Verify requests succeed against the loose file.
+        var caret = new LSP.Location
+        {
+            Range = new LSP.Range { Start = new LSP.Position { Character = 6, Line = 0 }, End = new LSP.Position { Character = 7, Line = 0 } },
+            Uri = looseFileUri,
+        };
+        var hover = await RunGetHoverAsync(testLspServer, caret).ConfigureAwait(false);
+        Assert.Contains("class A", hover.Contents.Third.Value);
+
+        // Update the workspace to contain the loose file.
+        var project = testLspServer.GetCurrentSolution().Projects.Single();
+        var documentInfo = DocumentInfo.Create(
+                DocumentId.CreateNewId(project.Id),
+                looseFileUri.AbsolutePath,
+                sourceCodeKind: SourceCodeKind.Regular,
+                loader: new TestTextLoader(source),
+                filePath: looseFileUri.AbsolutePath);
+        testLspServer.TestWorkspace.OnDocumentAdded(documentInfo);
```

Comment:
What component is responsible for this in VS? I assume the project system notifies us?

Reply:
It depends on what workspace it is.  For misc files in VS, we [listen to RDT events for document open](https://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs,83), then if we find it is a misc file, we [add it to the misc workspace](https://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs,267).

Files that are part of the project should be added to the VS workspace by the project system, [somewhere in here](https://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/VisualStudioProject.cs,1250).  And the RDT listener for the VS workspace only [opens the file](https://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs,122) that was already added by the project system.